pkg/brownfield/routing_rules.go

```diff
@@ -57,20 +57,73 @@ func (er ExistingResources) GetBlacklistedRoutingRules() ([]n.ApplicationGateway
 }
 
 // MergeRules merges list of lists of rules into a single list, maintaining uniqueness.
-func MergeRules(ruleBuckets ...[]n.ApplicationGatewayRequestRoutingRule) []n.ApplicationGatewayRequestRoutingRule {
-	uniq := make(rulesByName)
+func MergeRules(appGw *n.ApplicationGateway, ruleBuckets ...[]n.ApplicationGatewayRequestRoutingRule) []n.ApplicationGatewayRequestRoutingRule {
+	uniq := make(map[string]*n.ApplicationGatewayRequestRoutingRule)
 	for _, bucket := range ruleBuckets {
-		for _, rule := range bucket {
-			uniq[ruleName(*rule.Name)] = rule
+		for idx := range bucket {
+			rule := &bucket[idx]
+			// If two rules share the listener, we merge them. We keep the existing rule and merge the url path maps.
+			if existingRule, exists := uniq[*rule.HTTPListener.ID]; exists {
+				uniq[*rule.HTTPListener.ID] = DeepMergeRoutingRules(appGw, existingRule, rule)
+			} else {
+				uniq[*rule.HTTPListener.ID] = rule
+			}
 		}
 	}
 	var merged []n.ApplicationGatewayRequestRoutingRule
 	for _, rule := range uniq {
-		merged = append(merged, rule)
+		merged = append(merged, *rule)
 	}
 	return merged
 }
 
+// DeepMergeRoutingRules merges two routing rules by merging their pathRules
+func DeepMergeRoutingRules(appGw *n.ApplicationGateway, firstRoutingRule *n.ApplicationGatewayRequestRoutingRule, secondRoutingRule *n.ApplicationGatewayRequestRoutingRule) *n.ApplicationGatewayRequestRoutingRule {
+	if firstRoutingRule.RuleType == n.Basic && secondRoutingRule.RuleType == n.PathBasedRouting {
+		// Get the url path map of the second rule
+		glog.V(5).Infof("[brownfield] Merging basic rule %s with path based rule %s", *firstRoutingRule.Name, *secondRoutingRule.Name)
+		pathMap := LookupPathMap(appGw.URLPathMaps, secondRoutingRule.URLPathMap.ID)
+
+		// Replace the default values from the first rule
+		glog.V(5).Infof("[brownfield] Merging path map %s with rule %s", *pathMap.Name, *firstRoutingRule.Name)
+		MergePathMapsWithBasicRule(pathMap, firstRoutingRule)
+
+		// return second rule. Url path map is updated
+		return secondRoutingRule
+	} else if firstRoutingRule.RuleType == n.PathBasedRouting && secondRoutingRule.RuleType == n.Basic {
```

Comment:
`s/else/\n/`
and then you can just swap 1st/2nd and recurse: `return DeepMergeRoutingRules(appGw, second, first)`